python/cog/server/worker.py

```diff
@@ -373,32 +373,35 @@ def run(self) -> None:
         # Initially, we ignore SIGUSR1.
         signal.signal(signal.SIGUSR1, signal.SIG_IGN)
 
-        async_redirector = AsyncStreamRedirector(
-            callback=self._stream_write_hook,
-            tee=self._tee_output,
-        )
-
-        with async_redirector:
-            self._setup(async_redirector)
+        self._assign_predictor()
 
         # If setup didn't set the predictor, we're done here.
         if not self._predictor:
             return
 
         predict = get_predict(self._predictor)
         if inspect.iscoroutinefunction(predict) or inspect.isasyncgenfunction(predict):
+            async_redirector = AsyncStreamRedirector(
+                callback=self._stream_write_hook,
+                tee=self._tee_output,
+            )
+            with async_redirector:
+                self._setup(async_redirector)
             asyncio.run(self._aloop(predict, async_redirector))
```

Comment:
I think we still want `_setup` and `_aloop` to be inside the same `with` statement, just like the sync case below, so we're not exiting and re-entering the context manager

Reply:
Good point. I've updated the code to remove the additional `with redirector:` in the `_loop` and `_aloop`.